File: apps/painel-locacao/src/actions/locacoes.ts
'use server'

import { revalidarTelas } from '@/lib/revalidar'
import { z } from 'zod'
import { differenceInCalendarDays } from 'date-fns'
import { prisma } from '@/lib/prisma'
import { MOVIMENTACAO } from '@/lib/dominio/constantes'
import { dataBR } from '@/lib/dominio/formato'
import { obterLocacao } from '@/queries/locacoes'
import { exigirLancamento } from '@/lib/auth'

export type Resultado<T = void> = { ok: true; dados: T } | { ok: false; erro: string }

export async function carregarLocacao(id: string) {
  await exigirLancamento()
  return obterLocacao(id)
}

const esquemaLocacao = z.object({
  obraId: z.string().min(1, 'Selecione a obra.'),
  descricao: z.string().trim().min(2, 'Informe o equipamento.'),
  trCodigo: z.string().trim().optional(),
  fornecedorId: z.string().optional(),
  quantidade: z.coerce.number().int().min(1).default(1),
  dataInicio: z.coerce.date({ message: 'Data de início inválida.' }),
  dataFim: z.coerce.date({ message: 'Data de fim inválida.' }),
  valorItem: z.coerce.number().nonnegative().optional(),
  observacoes: z.string().trim().optional(),
})

export async function criarLocacao(entrada: unknown): Promise<Resultado<{ id: string }>> {
  await exigirLancamento()
  const parsed = esquemaLocacao.safeParse(entrada)
  if (!parsed.success) {
    return { ok: false, erro: parsed.error.issues.map((i) => i.message).join(' ') }
  }
  const d = parsed.data
  if (d.dataFim < d.dataInicio) return { ok: false, erro: 'A data de fim é anterior à de início.' }

  try {
    const locacao = await prisma.locacao.create({
      data: {
        obraId: d.obraId,
        descricao: d.descricao.toUpperCase(),
        trCodigo: d.trCodigo || null,
        fornecedorId: d.fornecedorId || null,
        quantidade: d.quantidade,
        dataInicio: d.dataInicio,
        dataFim: d.dataFim,
        valorItem: d.valorItem ?? null,
        observacoes: d.observacoes || null,
        movimentacoes: {
          create: {
            tipo: MOVIMENTACAO.REGISTRO,
            descricaoHumana: `Registrada de ${dataBR(d.dataInicio)} a ${dataBR(d.dataFim)}`,
          },
        },
      },
    })
    revalidarTelas('/', '/locacoes')
    return { ok: true, dados: { id: locacao.id } }
  } catch (e) {
    return { ok: false, erro: e instanceof Error ? e.message : 'Falha ao registrar' }
  }
}

export async function editarLocacao(id: string, entrada: unknown): Promise<Resultado> {
  await exigirLancamento()
  const parsed = esquemaLocacao.partial().safeParse(entrada)
  if (!parsed.success) {
    return { ok: false, erro: parsed.error.issues.map((i) => i.message).join(' ') }
  }
  try {
    const antes = await prisma.locacao.findUnique({ where: { id } })
    if (!antes) return { ok: false, erro: 'Locação não encontrada.' }

    const d = parsed.data

    // Valida a data contra o que FICARÁ gravado, não só contra o que veio no formulário.
    // Como a edição é parcial, alterar só o início pode invertê-lo em relação ao fim já
    // existente — e criar e transferir barram isso, mas editar não barrava. Uma locação
    // com fim antes do início produz duração negativa: o drawer mostra "Duração -5 dias"
    // e o Excel sai com período negativo mas valor cheio, ou seja, custo errado sem aviso.
    const inicioFinal = d.dataInicio ?? antes.dataInicio
    const fimFinal = d.dataFim ?? antes.dataFim
    if (inicioFinal && fimFinal && fimFinal < inicioFinal) {
      return { ok: false, erro: 'A data de fim ficaria anterior à de início.' }
    }

    await prisma.locacao.update({
      where: { id },
      data: {
        ...(d.obraId && { obraId: d.obraId }),
        ...(d.descricao && { descricao: d.descricao.toUpperCase() }),
        ...(d.trCodigo !== undefined && { trCodigo: d.trCodigo || null }),
        ...(d.fornecedorId !== undefined && { fornecedorId: d.fornecedorId || null }),
        ...(d.quantidade !== undefined && { quantidade: d.quantidade }),
        ...(d.dataInicio && { dataInicio: d.dataInicio }),
        ...(d.dataFim && { dataFim: d.dataFim }),
        ...(d.valorItem !== undefined && { valorItem: d.valorItem }),
        ...(d.observacoes !== undefined && { observacoes: d.observacoes || null }),
        movimentacoes: {
          create: {
            tipo: MOVIMENTACAO.EDICAO,
            descricaoHumana: 'Dados editados',
            payloadAntes: JSON.stringify(antes),
            payloadDepois: JSON.stringify(d),
          },
        },
      },
    })
    revalidarTelas('/', '/locacoes')
    return { ok: true, dados: undefined }
  } catch (e) {
    return { ok: false, erro: e instanceof Error ? e.message : 'Falha ao editar' }
  }
}

export async function renovarLocacao(id: string, diasExtras?: number): Promise<Resultado<{ novaData: Date }>> {
  await exigirLancamento()
  try {
    const l = await prisma.locacao.findUnique({ where: { id } })
    if (!l) return { ok: false, erro: 'Locação não encontrada.' }
    if (l.devolvidaEm) return { ok: false, erro: 'Locação já devolvida — não pode ser renovada.' }
    if (!l.dataInicio || !l.dataFim) return { ok: false, erro: 'Locação sem datas — edite antes de renovar.' }

    const dias = diasExtras ?? differenceInCalendarDays(l.dataFim, l.dataInicio)
    if (dias <= 0) return { ok: false, erro: 'Período de renovação inválido.' }

    const novaData = new Date(l.dataFim)
    novaData.setUTCDate(novaData.getUTCDate() + dias)

    await prisma.locacao.update({
      where: { id },
      data: {
        dataFim: novaData,
        movimentacoes: {
          create: {
            tipo: MOVIMENTACAO.RENOVACAO,
            descricaoHumana: `Renovada por ${dias} dias — novo fim ${dataBR(novaData)}`,
            payloadAntes: JSON.stringify({ dataFim: l.dataFim }),
            payloadDepois: JSON.stringify({ dataFim: novaData }),
          },
        },
      },
    })
    revalidarTelas('/', '/locacoes')
    return { ok: true, dados: { novaData } }
  } catch (e) {
    return { ok: false, erro: e instanceof Error ? e.message : 'Falha ao renovar' }
  }
}

/**
 * Preenche `devolvidaEm` e registra a movimentação. A linha NÃO é apagada nem recriada:
 * `dataInicio` fica intacta, então o tempo real que o equipamento passou na obra continua
 * legível. O app antigo apagava a linha e a recriava no bloco de devoluções com a data de
 * início sobrescrita — nas 63 devoluções importadas da planilha, início = fim, e o período
 * real já se perdeu. É esse apagamento que esta função existe para impedir.
 */
export async function devolverLocacao(id: string, dataDevolucao: Date, motivo?: string): Promise<Resultado> {
  await exigirLancamento()
  try {
    const l = await prisma.locacao.findUnique({ where: { id } })
    if (!l) return { ok: false, erro: 'Locação não encontrada.' }
    if (l.devolvidaEm) return { ok: false, erro: 'Locação já devolvida.' }

    await prisma.locacao.update({
      where: { id },
      data: {
        devolvidaEm: dataDevolucao,
        movimentacoes: {
          create: {
            tipo: MOVIMENTACAO.DEVOLUCAO,
            descricaoHumana:
              `Devolvida em ${dataBR(dataDevolucao)}` +
              (l.dataInicio ? ` — permaneceu ${differenceInCalendarDays(dataDevolucao, l.dataInicio)} dias na obra` : '') +
              (motivo ? ` · ${motivo}` : ''),
          },
        },
      },
    })
    revalidarTelas('/', '/locacoes')
    return { ok: true, dados: undefined }
  } catch (e) {
    return { ok: false, erro: e instanceof Error ? e.message : 'Falha ao devolver' }
  }
}

export async function transferirLocacao(
  id: string, obraDestinoId: string, dataInicio: Date, dataFim: Date, motivo?: string
): Promise<Resultado> {
  await exigirLancamento()
  try {
    const l = await prisma.locacao.findUnique({ where: { id }, include: { obra: true } })
    if (!l) return { ok: false, erro: 'Locação não encontrada.' }
    if (l.devolvidaEm) return { ok: false, erro: 'Locação devolvida não pode ser transferida.' }
    if (l.obraId === obraDestinoId) return { ok: false, erro: 'A obra de destino é a mesma da origem.' }
    if (dataFim < dataInicio) return { ok: false, erro: 'A data de fim é anterior à de início.' }

    const destino = await prisma.obra.findUnique({ where: { id: obraDestinoId } })
    if (!destino) return { ok: false, erro: 'Obra de destino não encontrada.' }

    await prisma.locacao.update({
      where: { id },
      data: {
        obraId: obraDestinoId,
        dataInicio, dataFim,
        obraAConfirmar: false,
        movimentacoes: {
          create: {
            tipo: MOVIMENTACAO.TRANSFERENCIA,
            descricaoHumana:
              `Transferida de ${l.obra.codigo} para ${destino.codigo}` +
              ` — novo período ${dataBR(dataInicio)} a ${dataBR(dataFim)}` +
              (motivo ? ` · ${motivo}` : ''),
            payloadAntes: JSON.stringify({ obra: l.obra.codigo, dataInicio: l.dataInicio, dataFim: l.dataFim }),
            payloadDepois: JSON.stringify({ obra: destino.codigo, dataInicio, dataFim }),
          },
        },
      },
    })
    revalidarTelas('/', '/locacoes')
    return { ok: true, dados: undefined }
  } catch (e) {
    return { ok: false, erro: e instanceof Error ? e.message : 'Falha ao transferir' }
  }
}

export async function reclassificarEmLote(ids: string[], obraDestinoId: string): Promise<Resultado<{ movidas: number }>> {
  await exigirLancamento()
  if (!ids.length) return { ok: false, erro: 'Nenhum item selecionado.' }
  try {
    const destino = await prisma.obra.findUnique({ where: { id: obraDestinoId } })
    if (!destino) return { ok: false, erro: 'Obra de destino não encontrada.' }

    await prisma.$transaction([
      prisma.locacao.updateMany({
        where: { id: { in: ids } },
        data: { obraId: obraDestinoId, obraAConfirmar: false },
      }),
      prisma.movimentacao.createMany({
        data: ids.map((locacaoId) => ({
          locacaoId,
          tipo: MOVIMENTACAO.RECLASSIFICACAO,
          descricaoHumana: `Obra confirmada como ${destino.codigo}`,
        })),
      }),
    ])
    revalidarTelas('/', '/locacoes')
    return { ok: true, dados: { movidas: ids.length } }
  } catch (e) {
    return { ok: false, erro: e instanceof Error ? e.message : 'Falha ao reclassificar' }
  }
}

File: migracao-go/internal/application/painel/locacoes.go
package painel

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	dominio "siqueiracampos/servidor/internal/domain/painel"
)

type GerenciadorLocacoes struct {
	Locacoes dominio.LocacaoRepositorio
	Obras    dominio.ObraRepositorio
}

type EntradaLocacao struct {
	ObraID       string
	Descricao    string
	TrCodigo     string
	FornecedorID string
	Quantidade   int
	DataInicio   *time.Time
	DataFim      *time.Time
	ValorItem    *float64
	Observacoes  string
}

func validarEntradaLocacao(e EntradaLocacao, exigirCampos bool) (erros []string) {
	if exigirCampos && strings.TrimSpace(e.ObraID) == "" {
		erros = append(erros, "Selecione a obra.")
	}
	if exigirCampos && len(strings.TrimSpace(e.Descricao)) < 2 {
		erros = append(erros, "Informe o equipamento.")
	}
	if exigirCampos && e.DataInicio == nil {
		erros = append(erros, "Data de início inválida.")
	}
	if exigirCampos && e.DataFim == nil {
		erros = append(erros, "Data de fim inválida.")
	}
	return
}

// Criar espelha `criarLocacao` — ver COMPORTAMENTO.md §4.3.
func (g *GerenciadorLocacoes) Criar(ctx context.Context, e EntradaLocacao) (*dominio.Locacao, error) {
	if erros := validarEntradaLocacao(e, true); len(erros) > 0 {
		return nil, erroValidacao(erros...)
	}
	if e.DataFim.Before(*e.DataInicio) {
		return nil, erroValidacao("A data de fim é anterior à de início.")
	}

	quantidade := e.Quantidade
	if quantidade < 1 {
		quantidade = 1
	}

	l := &dominio.Locacao{
		ObraID: e.ObraID, Descricao: strings.ToUpper(strings.TrimSpace(e.Descricao)),
		TrCodigo: ponteiro(e.TrCodigo), Quantidade: quantidade, Estado: dominio.EstadoOK,
		Observacoes: ponteiro(e.Observacoes), DataInicio: e.DataInicio, DataFim: e.DataFim,
		ValorItem: e.ValorItem, FornecedorID: ponteiro(e.FornecedorID),
	}
	mov := dominio.Movimentacao{
		Tipo:            dominio.MovRegistro,
		DescricaoHumana: fmt.Sprintf("Registrada de %s a %s", dominio.DataBR(e.DataInicio), dominio.DataBR(e.DataFim)),
	}
	if err := g.Locacoes.Criar(ctx, l, mov); err != nil {
		return nil, err
	}
	return l, nil
}

// EntradaEdicao é parcial — campo nil/vazio significa "não alterar", espelhando o schema
// `.partial()` de editarLocacao em Next.js.
type EntradaEdicao struct {
	ObraID           *string
	Descricao        *string
	TrCodigo         *string
	FornecedorID     *string
	Quantidade       *int
	DataInicio       *time.Time
	DataFim          *time.Time
	ValorItem        *float64
	AlterarValorItem bool // false = não mexe no campo; true = grava ValorItem (mesmo nil, pra limpar)
	Observacoes      *string
}

// Editar espelha `editarLocacao` — ver COMPORTAMENTO.md §4.3: valida a data contra o que
// FICARÁ gravado (mistura do que veio com o que já existia), não só contra o formulário.
func (g *GerenciadorLocacoes) Editar(ctx context.Context, id string, e EntradaEdicao) error {
	antes, err := g.Locacoes.BuscarPorID(ctx, id)
	if err != nil {
		return err
	}
	if antes == nil {
		return erroValidacao("Locação não encontrada.")
	}

	depois := *antes
	antesJSON, _ := json.Marshal(camposEditaveis(antes))

	if e.ObraID != nil {
		depois.ObraID = *e.ObraID
	}
	if e.Descricao != nil {
		d := strings.TrimSpace(*e.Descricao)
		if len(d) > 0 {
			depois.Descricao = strings.ToUpper(d)
		}
	}
	if e.TrCodigo != nil {
		depois.TrCodigo = ponteiro(*e.TrCodigo)
	}
	if e.FornecedorID != nil {
		depois.FornecedorID = ponteiro(*e.FornecedorID)
	}
	if e.Quantidade != nil {
		depois.Quantidade = *e.Quantidade
	}
	if e.DataInicio != nil {
		depois.DataInicio = e.DataInicio
	}
	if e.DataFim != nil {
		depois.DataFim = e.DataFim
	}
	if e.AlterarValorItem {
		depois.ValorItem = e.ValorItem
	}
	if e.Observacoes != nil {
		depois.Observacoes = ponteiro(*e.Observacoes)
	}

	if depois.DataInicio != nil && depois.DataFim != nil && depois.DataFim.Before(*depois.DataInicio) {
		return erroValidacao("A data de fim ficaria anterior à de início.")
	}

	depoisJSON, _ := json.Marshal(camposEditaveis(&depois))
	antesStr, depoisStr := string(antesJSON), string(depoisJSON)
	mov := dominio.Movimentacao{Tipo: dominio.MovEdicao, DescricaoHumana: "Dados editados", PayloadAntes: &antesStr, PayloadDepois: &depoisStr}

	return g.Locacoes.Atualizar(ctx, &depois, mov)
}

func camposEditaveis(l *dominio.Locacao) map[string]any {
	return map[string]any{
		"obraId": l.ObraID, "descricao": l.Descricao, "trCodigo": l.TrCodigo,
		"fornecedorId": l.FornecedorID, "quantidade": l.Quantidade,
		"dataInicio": dominio.DataBR(l.DataInicio), "dataFim": dominio.DataBR(l.DataFim),
		"valorItem": l.ValorItem, "observacoes": l.Observacoes,
	}
}

// Renovar espelha `renovarLocacao` — ver COMPORTAMENTO.md §4.3. diasExtras nil usa a
// duração atual da locação (fim-início) como o tanto a acrescentar.
func (g *GerenciadorLocacoes) Renovar(ctx context.Context, id string, diasExtras *int) (time.Time, error) {
	l, err := g.Locacoes.BuscarPorID(ctx, id)
	if err != nil {
		return time.Time{}, err
	}
	if l == nil {
		return time.Time{}, erroValidacao("Locação não encontrada.")
	}
	if l.DevolvidaEm != nil {
		return time.Time{}, erroValidacao("Locação já devolvida — não pode ser renovada.")
	}
	if l.DataInicio == nil || l.DataFim == nil {
		return time.Time{}, erroValidacao("Locação sem datas — edite antes de renovar.")
	}

	dias := dominio.DuracaoEmDias(l.DataInicio, l.DataFim)
	if diasExtras != nil {
		dias = *diasExtras
	}
	if dias <= 0 {
		return time.Time{}, erroValidacao("Período de renovação inválido.")
	}

	novaData := l.DataFim.UTC().AddDate(0, 0, dias)
	antesJSON, _ := json.Marshal(map[string]any{"dataFim": dominio.DataBR(l.DataFim)})
	depoisJSON, _ := json.Marshal(map[string]any{"dataFim": dominio.DataBR(&novaData)})
	antesStr, depoisStr := string(antesJSON), string(depoisJSON)

	l.DataFim = &novaData
	mov := dominio.Movimentacao{
		Tipo:            dominio.MovRenovacao,
		DescricaoHumana: fmt.Sprintf("Renovada por %d dias — novo fim %s", dias, dominio.DataBR(&novaData)),
		PayloadAntes:    &antesStr, PayloadDepois: &depoisStr,
	}
	if err := g.Locacoes.Atualizar(ctx, l, mov); err != nil {
		return time.Time{}, err
	}
	return novaData, nil
}

// Devolver espelha `devolverLocacao` — ver COMPORTAMENTO.md §4.3: a linha NUNCA é apagada
// nem recriada, só `devolvidaEm` é preenchida. É o comportamento mais importante do módulo.
func (g *GerenciadorLocacoes) Devolver(ctx context.Context, id string, dataDevolucao time.Time, motivo string) error {
	l, err := g.Locacoes.BuscarPorID(ctx, id)
	if err != nil {
		return err
	}
	if l == nil {
		return erroValidacao("Locação não encontrada.")
	}
	if l.DevolvidaEm != nil {
		return erroValidacao("Locação já devolvida.")
	}

	descricao := fmt.Sprintf("Devolvida em %s", dominio.DataBR(&dataDevolucao))
	if l.DataInicio != nil {
		permanencia := dominio.DuracaoEmDias(l.DataInicio, &dataDevolucao)
		descricao += fmt.Sprintf(" — permaneceu %d dias na obra", permanencia)
	}
	if motivo = strings.TrimSpace(motivo); motivo != "" {
		descricao += " · " + motivo
	}

	l.DevolvidaEm = &dataDevolucao
	mov := dominio.Movimentacao{Tipo: dominio.MovDevolucao, DescricaoHumana: descricao}
	return g.Locacoes.Atualizar(ctx, l, mov)
}

// Transferir espelha `transferirLocacao` — ver COMPORTAMENTO.md §4.3.
func (g *GerenciadorLocacoes) Transferir(ctx context.Context, id, obraDestinoID string, dataInicio, dataFim time.Time, motivo string) error {
	l, err := g.Locacoes.BuscarPorID(ctx, id)
	if err != nil {
		return err
	}
	if l == nil {
		return erroValidacao("Locação não encontrada.")
	}
	if l.DevolvidaEm != nil {
		return erroValidacao("Locação devolvida não pode ser transferida.")
	}
	if l.ObraID == obraDestinoID {
		return erroValidacao("A obra de destino é a mesma da origem.")
	}
	if dataFim.Before(dataInicio) {
		return erroValidacao("A data de fim é anterior à de início.")
	}

	destino, err := g.Obras.BuscarPorID(ctx, obraDestinoID)
	if err != nil {
		return err
	}
	if destino == nil {
		return erroValidacao("Obra de destino não encontrada.")
	}

	descricao := fmt.Sprintf("Transferida de %s para %s — novo período %s a %s",
		l.ObraCodigo, destino.Codigo, dominio.DataBR(&dataInicio), dominio.DataBR(&dataFim))
	if motivo = strings.TrimSpace(motivo); motivo != "" {
		descricao += " · " + motivo
	}
	antesJSON, _ := json.Marshal(map[string]any{"obra": l.ObraCodigo, "dataInicio": dominio.DataBR(l.DataInicio), "dataFim": dominio.DataBR(l.DataFim)})
	depoisJSON, _ := json.Marshal(map[string]any{"obra": destino.Codigo, "dataInicio": dominio.DataBR(&dataInicio), "dataFim": dominio.DataBR(&dataFim)})
	antesStr, depoisStr := string(antesJSON), string(depoisJSON)

	l.ObraID = obraDestinoID
	l.DataInicio, l.DataFim = &dataInicio, &dataFim
	l.ObraAConfirmar = false
	mov := dominio.Movimentacao{Tipo: dominio.MovTransferencia, DescricaoHumana: descricao, PayloadAntes: &antesStr, PayloadDepois: &depoisStr}
	return g.Locacoes.Atualizar(ctx, l, mov)
}

// ReclassificarEmLote espelha `reclassificarEmLote` — ver COMPORTAMENTO.md §4.3.
func (g *GerenciadorLocacoes) ReclassificarEmLote(ctx context.Context, ids []string, obraDestinoID string) (int, error) {
	if len(ids) == 0 {
		return 0, erroValidacao("Nenhum item selecionado.")
	}
	destino, err := g.Obras.BuscarPorID(ctx, obraDestinoID)
	if err != nil {
		return 0, err
	}
	if destino == nil {
		return 0, erroValidacao("Obra de destino não encontrada.")
	}
	if err := g.Locacoes.ReclassificarEmLote(ctx, ids, obraDestinoID, destino.Codigo); err != nil {
		return 0, err
	}
	return len(ids), nil
}
